wgengine/netstack/netstack.go

```diff
@@ -793,3 +796,56 @@ func stringifyTEI(tei stack.TransportEndpointID) string {
 func ipPortOfNetstackAddr(a tcpip.Address, port uint16) (ipp netaddr.IPPort, ok bool) {
 	return netaddr.FromStdAddr(net.IP(a), int(port), "") // TODO(bradfitz): can do without allocs
 }
+
+// removeECNBits modifies p to remove ECN bits, if any.
+func removeECNBits(p *packet.Parsed) {
+	if p.IPVersion == 0 {
+		return
+	}
+	if p.IPProto != ipproto.TCP {
+		// TODO(bradfitz): handle non-TCP too? for now only trying to
+		// fix the Issue 2642 problem.
+		return
+	}
+	if p.TCPFlags&(packet.TCPCWR|packet.TCPECNEcho) == 0 {
+		// Nothing to do.
+		return
+	}
+
+	// Clear flags.
```

Comment:
Feels a bit dicey to be reaching into parsed packets this far from where they're parsed, but I'll hold my nose since this is for debugging.

For more future robustness, make it a method on packet.Parsed, so the mangling of raw bytes is in the same place as the parsing code?

Reply:
This isn't just for debugging; this is meant to be the fix.

> For more future robustness, make it a method on packet.Parsed, so the mangling of raw bytes is in the same place as the parsing code?

So that's how I started, but then I got lost in horrible checksum calculations and debating pulling out & rewriting gvisor's checksumming (because I hate its string-based tcpip.Address type) and then I moved it to wgengine/netstack so I could just depend on the gvisor packages, but then ultimately I ended up not doing that and finding out how to incrementally update the checksum.

But now that I've done that, I can move it back to net/packet.

Will do.